### ISSUE TITLE
Access to intermediate HASH data is need to support ISO7816-8 "5.5 HASH operation"

### ISSUE THREAD

**Issue by dengert** [openssl/openssl#17688]
EVP_MD_CTX_get0_md_data is not deprecated in 3.0.1 but it returns a pointer to hash structures which are deprecated, such as SHA_CTX and SHA256_CTX in our case. I believe this is an oversight. 

ISO/IEC 7816-8 "Identification cards — Integrated circuit cards — Part 8: Commands for security operations"  "5.5 HASH operation" says:
"The HASH operation initiates the computation of a hash-code by performing:
 either the complete computation inside the card or
 **a partial computation inside the card (e.g. last round of hashing).**"

The above requires the intermediate hash data to be accessible to be sent to the card so the card can do last round of hashing. 

OpenSC has used OpenSSL hash routines for SHA1 and SHA256 with cards that can do: "a partial computation inside the card (e.g. last round of hashing)."  in previous version but needs a way to get access to at least the SHA256_CTX in 3.0.1.

A proposed patch is to copy the SHA_CTX and SHA256_CTX definition and other defines is here:
https://github.com/OpenSC/OpenSC/pull/2506/commits/1f3df786738431fa00fab23a6a830f32eb3e24e0
But is based on the structures not changing in future versions of OpenSSL.

Can the intermediate hash data for at least SHA256 not be deprecated or made available in some other way? 

Thanks.

**Comment by dengert** [comment]
I proposed solution is to move two `#ifndef OPENSSL_NO_DEPRECATED_3_0` statements in `sha.h` This would then leave the SHA_CTX and SHA256_CTX defined in future versions. 

[OpenSSL-3-sha.diff.txt](https://github.com/openssl/openssl/files/8054153/OpenSSL-3-sha.diff.txt)

**Comment by dengert** [comment]
Yes. For the SHA256 case the pseudo code does:
``` 
        md = EVP_get_digestbyname("SHA256");
	mdctx = EVP_MD_CTX_new();
	EVP_DigestInit_ex(mdctx, md, NULL);.       
       md_data =  EVP_MD_CTX_get0_md_data(mdctx); /* gets pointer to SHA256  parameters */
       EVP_DigestUpdate(mdctx, in, in_len);
```
https://github.com/OpenSC/OpenSC/commit/1f3df786738431fa00fab23a6a830f32eb3e24e0  is a patch to OpenSC to update the code to work with OpenSSL 3.0 as well as 1.1.1 and LibreSSL.

From the `SHA256_CTX` the code uses:  `SHA_LONG h[8];  SHA_LONG Nl, Nh;` and `unsigned int num;`
(the code could also use `SHA_LONG data[SHA_LBLOCK];` but the code calculates from its input.) 

This use of the `SHA256_CTX` data is in the patch at: https://github.com/OpenSC/OpenSC/commit/1f3df786738431fa00fab23a6a830f32eb3e24e0#diff-887c1648a5462ce3cbbcef62689ecd3c7fe57fa0e748ee713a36bd1dfce33cfeR3324-R3346

The code does go on to call `EVP_DigestFinal_ex(mdctx, out->hash, &md_out_len);` for debugging and to verify that the hash created by the card matched the software hash. The security requirement is the smart card will create the last round of the hash.

I did not write the original code and do not have the smart card to test it. But I did write the patch. OpenSC developers are concerned on how to handle this smart card if this data will not be available in future versions of OpenSSL. 

It is also not clear if `SHA_LONG Nl, Nh;  and unsigned int num;` are really needed or could be derived from the size of the input data and HASH.  But the `SHA_LONG h[8];` are needed to seed the card with the intermediate data.

**Comment by t-j-h** [comment]
> So, if I understand correctly, what you want is to be able to get a blob of some sort that represents the current state, and to be able to feed that same blob back to that same algorithm (card) at a later time to continue the digest operation. Correct?

No - that is not what is being asked for. This is not save-state, restore-state to the same implementation. It is extract the internal state from the algorithm in one implementation and pass that internal state off to another implementation - i.e. one implementation is openssl (off card) and one is something entirely different (on card) and we are passing state between the two.

The "solution" should not be to expose internal structures - it should be to have a defined way to export in a portable format the state without using the internal OpenSSL implementation structure.

Note that this sort of usage is strictly not permitted in a FIPS140 context.

**Comment by dengert** [comment]
> It is extract the internal state from the algorithm in one implementation and pass that internal state off to another implementation 

Yes, that is what I am asking for. OpenSSL does the first part, and the card does the final round. 

> The "solution" should not be to expose internal structures - it should be to have a defined way to export in a portable format the state without using the internal OpenSSL implementation structure.

Then what is the purpose of `EVP_MD_CTX_get0_md_data`? 
It is not deprecated in 3.0.1 and returns a pointer to a SHA256_CTX. when using "SHA256" and SHA_CTX when using "SHA1". (gdb of a test program shows the "H0" matches the initial vector for SHA256 or SHA1.) 

Call it what you want, the intermediate SHA256 "H" numbers are needed and could be passed as some blob. They just happen to be the first part of the SHA256_CTX after a call to EVP_DigestUpdate. The `Nl` and `Nh` are also used by the implementation, but (I suspect) they can be derived from the input data. The OpenSC software then send any remaining data and values need for the card to do the "5.0 PREPROCESSING", "5.1 Padding the Message" and "5.1.1 Padding the Message" as found in" https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.180-4.pdf 

I would like to suggest that the "solution" use a `SHA256_CTX` like structure even if the internal OpenSSL implementation structure is something else.  

> Note that this sort of usage is strictly not permitted in a FIPS140 context.

That may be the case, but in this case we are relying on ISO/IEC 7816-8 "Identification cards — Integrated circuit cards — Part 8: Commands for security operations" "5.5 HASH operation" calls for: "a partial computation inside the card (e.g. last round of hashing)."

**Comment by t8m** [comment]
> Then what is the purpose of `EVP_MD_CTX_get0_md_data`? It is not deprecated in 3.0.1 and returns a pointer to a SHA256_CTX. when using "SHA256" and SHA_CTX when using "SHA1". (gdb of a test program shows the "H0" matches the initial vector for SHA256 or SHA1.)

It was not deprecated by mistake. It should have been. We cannot in any reasonable way support its operation with provider implementations.

**Comment by levitte** [comment]
> The "solution" should not be to expose internal structures - it should be to have a defined way to export in a portable format the state without using the internal OpenSSL implementation structure.

So basically, a "blob" with the well defined structure and contents...  I agree with that, and thought I wrote my previous comment to include that possibility.  Looking back, I can see that wasn't quite clear.

> Note that this sort of usage is strictly not permitted in a FIPS140 context.

... and considering that there is no well known structure for this sort of thing, it's currently a bit hard to do that anyway.  Of course, nothing stops us from inventing something (and oh joy, specify it in modern ASN.1?), but that's quite an effort...

**Comment by mattcaswell** [comment]
Not sure why it has to a blob. Why can't we just have params for the individual fields?

**Comment by levitte** [comment]
> Not sure why it has to a blob. Why can't we just have params for the individual fields?

No reason why not ...  "blob" was metaphorical rather than specific

**Comment by dengert** [comment]
It looks like in 3.0.1 `EVP_MD_CTX_get0_md_data` is not deprecated, but does return NULL. So my proposed fix from above is not complete. 

Since the OpenSC code has control of the calling of EVP_DigestUpdate, it can send a multiple the SHA input so that no data is held over for the  "final" round, thus only the array of "H" values are needed in the "blob" and not a full structure. 

If this can not be easily done in OpenSSL, OpenSC has some other options like write a custom provider that can return the data. Or OpenSC could use some other implementation of SHA256 for use for this one card that supports the hash on the card. Or not support this one card if using OpenSSL.

The use of the hash on the card is used to create a signature. If I understand your FIPS140 argument, the verification of the signature is where the hash should not expose the internal data or accept partially provided input. But in our case the creation of the hash would only provide a valid signature if both the software and final round on card results have done the has correctly.

**Comment by levitte** [comment]
> Then what is the purpose of `EVP_MD_CTX_get0_md_data`?

The purpose is for a (now legacy) backend implementation to access the backend specific data from a `EVP_MD_CTX`.  It's now legacy, because this isn't at all how things work with providers, which is why you're getting NULL, 'cause providers keep those structures very much more to themselves.

Trying to use this data for higher level purposes is quite a hack, and depends on specific backends; for example, an engine could implement SHA1 or SHA256 using a completely different structure for its own contexts...  you cannot assume that it's a `SHA_CTX` or `SHA256_CTX` as OpenSSL used to define them.
In other words, using the contents of what is returned from `EVP_MD_CTX_get0_md_data` is undefined in all places except the backend that governs those contents.

That being said, previous comments still apply.  It can be made possible to extract relevant bits of data in some sort of supported manner.  This would be a new feature.

**Comment by DemiMarie** [comment]
This is also needed by libtpms: https://github.com/stefanberger/libtpms/issues/356#issuecomment-1447151135
